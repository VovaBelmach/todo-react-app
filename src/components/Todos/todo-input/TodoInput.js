import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { addTodo } from "../../../stores/todosSlice";
import Card from "../../UI/Card/Card";
import CustomTodoInput from "../../UI/CustomTodoInput/CustomTodoInput";
import styles from "./TodoInput.module.css";
import {
  TODO_INPUT_EMPTY_ERROR_MESSAGE,
  TODO_INPUT_PLACEHOLDER_TEXT,
} from "../../../constants";

const TodoInput = () => {
  const [enteredTodo, setEnteredTodo] = useState("");
  const [error, setError] = useState("");
  const dispatch = useDispatch();
  const todoInputName = "todo-input";

  const onSubmitHandler = (event) => {
    event.preventDefault();

    if (enteredTodo.trim().length === 0) {
      setError(TODO_INPUT_EMPTY_ERROR_MESSAGE);
      return;
    }

    dispatch(
      addTodo({
        description: enteredTodo,
      })
    );

    setEnteredTodo("");
  };

  const todoChangeHandler = (event) => {
    setEnteredTodo(event.target.value);
    setError("");
  };

  return (
    <Card className={styles["todo-input"]}>
      <form aria-label="form" onSubmit={onSubmitHandler}>
        <CustomTodoInput
          name={todoInputName}
          value={enteredTodo}
          placeholder={TODO_INPUT_PLACEHOLDER_TEXT}
          onChangeHandler={todoChangeHandler}
        />
      </form>
      <span className={styles["error-message"]}>{error}</span>
    </Card>
  );
};

export default TodoInput;
