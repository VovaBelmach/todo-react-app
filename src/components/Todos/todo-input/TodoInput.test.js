import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../../../stores/configureStore";
import TodoInput from "./TodoInput";

describe("TodoInput component", () => {
  it("should render the TodoInput component", () => {
    render(
      <Provider store={store}>
        <TodoInput />
      </Provider>
    );
    const inputElement = screen.getByPlaceholderText("What should be done?");
    expect(inputElement).toBeInTheDocument();
  });

  it("should show an error message when submitting an empty todo", () => {
    render(
      <Provider store={store}>
        <TodoInput />
      </Provider>
    );

    const formElement = screen.getByRole("form");
    fireEvent.submit(formElement);
    const errorMessage = screen.getByText(
      "Oops! You cannot create an empty todo. Please provide 'What should be done?' in the section above."
    );

    expect(errorMessage).toBeInTheDocument();
  });

  it("should dispatch an action when submitting a non-empty todo", () => {
    render(
      <Provider store={store}>
        <TodoInput />
      </Provider>
    );

    const inputElement = screen.getByPlaceholderText("What should be done?");
    const formElement = screen.getByRole("form");
    const enteredTodo = "Test todo";
    fireEvent.change(inputElement, { target: { value: enteredTodo } });
    fireEvent.submit(formElement);
    const state = store.getState();
    const todos = state.todos.todos;
    
    expect(todos).toHaveLength(1);
    expect(todos[0].description).toBe(enteredTodo);
  });
});
