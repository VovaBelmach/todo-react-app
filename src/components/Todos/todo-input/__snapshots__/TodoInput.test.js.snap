// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInput component should render correctly 1`] = `
<div
  className="card todo-input"
>
  <form
    aria-label="form"
    onSubmit={[Function]}
  >
    <label>
      <input
        className="checkbox"
        disabled={true}
        type="checkbox"
      />
    </label>
    <input
      name="todo-input"
      onChange={[Function]}
      placeholder="What should be done?"
      type="text"
      value=""
    />
  </form>
  <span
    className="error-message"
  />
</div>
`;
